libyara/modules/dotnet/dotnet.c

```diff
@@ -16,66 +16,1639 @@ limitations under the License.
+        ctx->tables->typespec.RowCount);
+
+    if (row_count > (0xFFFF >> 0x02))
+      interface_size = 4;
+
+    result->Class = read_index(&data, ctx->index_sizes->typedef_);
+    result->Interface = read_index(&data, interface_size);
+
+    return true;
+  }
+
+  return false;
+}
+
+static bool read_methoddef(
+    const CLASS_CONTEXT* ctx,
+    const uint8_t* data,
+    METHODDEF_ROW* result)
+{
+  uint32_t row_size = ctx->tables->methoddef.RowSize;
+
+  if (fits_in_pe(ctx->pe, data, row_size))
+  {
+    result->Rva = read_u32(&data);
+    result->ImplFlags = read_u16(&data);
+    result->Flags = read_u16(&data);
+    result->Name = read_index(&data, ctx->index_sizes->string);
+    result->Signature = read_index(&data, ctx->index_sizes->blob);
+    result->ParamList = read_index(&data, ctx->index_sizes->param);
+    return true;
+  }
+
+  return false;
+}
+
+static bool read_param(
+    const CLASS_CONTEXT* ctx,
+    const uint8_t* data,
+    PARAM_ROW* result)
+{
+  uint32_t row_size = ctx->tables->param.RowSize;
+
+  if (fits_in_pe(ctx->pe, data, row_size))
+  {
+    result->Flags = read_u16(&data);
+    result->Sequence = read_u16(&data);
+    result->Name = read_index(&data, ctx->index_sizes->string);
+    return true;
+  }
+
+  return false;
+}
+
+static bool read_genericparam(
+    const CLASS_CONTEXT* ctx,
+    const uint8_t* data,
+    GENERICPARAM_ROW* result)
+{
+  uint32_t row_size = ctx->tables->genericparam.RowSize;
+
+  if (fits_in_pe(ctx->pe, data, row_size))
+  {
+    uint32_t owner_idx_size = 2;
+    uint32_t row_count = max_rows(
+        2, ctx->tables->typedef_.RowCount, ctx->tables->methoddef.RowCount);
+
+    if (row_count > (0xFFFF >> 0x01))
+      owner_idx_size = 4;
+
+    result->Number = read_u16(&data);
+    result->Flags = read_u16(&data);
+    result->Owner = read_index(&data, owner_idx_size);
+    result->Name = read_index(&data, ctx->index_sizes->string);
+    return true;
+  }
+
+  return false;
+}
+
+static bool read_typespec(
+    const CLASS_CONTEXT* ctx,
+    const uint8_t* data,
+    TYPESPEC_ROW* result)
+{
+  uint32_t row_size = ctx->tables->typespec.RowSize;
+
+  if (fits_in_pe(ctx->pe, data, row_size))
+  {
+    result->Signature = read_index(&data, ctx->index_sizes->blob);
+    return true;
+  }
+
+  return false;
+}
+
+static bool read_nestedclass(
+    const CLASS_CONTEXT* ctx,
+    const uint8_t* data,
+    NESTEDCLASS_ROW* result)
+{
+  uint32_t row_size = ctx->tables->nestedclass.RowSize;
+
+  if (fits_in_pe(ctx->pe, data, row_size))
+  {
+    result->NestedClass = read_index(&data, ctx->index_sizes->typedef_);
+    result->EnclosingClass = read_index(&data, ctx->index_sizes->typedef_);
+    return true;
+  }
+
+  return false;
+}
+
+// ECMA-335 II.23.2 blob heap uses variable length encoding of integers
+static uint32_t read_blob_unsigned(const uint8_t** data, uint32_t* len)
+{
+  if (*len < 1)
+    return 0;
+
+  /* first byte is enough to decode the length
```

Comment:
Use `//` for all comments. When you want to comment whole functions or large sections of code it becomes harder if you have `/**/` in the code, as this types of comments can't be nested.